File: modules/brlx/native/filestream.java

class FileStream extends Stream{

	boolean Open( String path,String mode ){
	
		if( _file!=null ) return false;
	
		try{
			_file=new RandomAccessFile( path,mode );
		}catch( FileNotFoundException ex ){
			return false;
		}
		_off=_file.getFilePointer();
		_len=_file.length();
	}

	int Length(){
		return (int)_len;
	}
	
	int Offset(){
		return (int)_off;
	}
	
	int Seek( int offset ){
		_file.seek( offset );
		_off=_file.getFilePointer();
	
	}
		
	int Eof(){
		return (_off==_len) ? 1 : 0;
	}
	
	void Close(){
		if( _file==null ) return;
		_file=null;
		_off=0;
		_len=0;
	}
	
	int SkipBytes( int count ){
	
		if( _file==null ) return 0;
		
		return _file.skipBytes( count );
	}
	
	int ReadBytes( DataBuffer buffer,int count,int offset ){
	
		if( _file==null ) return 0;
		
		ByteBuffer buf=buffer.GetBuffer();
		
		if( buf.isDirect() ){
			byte[] arr=new byte[count];
			n=_file.read( arr,0,count );
			if( n<=0 ) return 0;
			buf.mark();
			buf.position( offset );
			buf.put( arr,0,n );
			buf.reset();
			return n;
		}else{
			int n=_file.read( buf.array(),buf.offset()+offset,count );
			return n>0 ? n : 0;
		}
	}
	
	int WriteBytes( DataBuffer buffer,int count,int offset ){
	
		if( _file==null ) return 0;
		
		ByteBuffer buf=buffer.GetByteBuffer();
		
		if( buf.isDirect() ){
			byte[] arr=new byte[count];
			buf.mark();
			buf.position( offset );
			buf.get( arr,0,count );
			buf.reset();
			int n=_file.write( arr,0,count );
			return n>0 ? n : 0;
		}else{
			int n=_file.write( buf.array(),buf.offset()+offset,count );
			return n>0 ? n : 0;
		}
	}

	RandromAccessFile _file;
	long _off,_len;
}

File: modules/brlx/native/filestream.cpp

// ***** filestream.h *****

class BBFileStream : public BBStream{
public:

	BBFileStream();
	~BBFileStream();

	int Length();
	int Offset();

	int Seek( int offset );
	
	int Eof();
	void Close();
	int SkipBytes( int count );
	int ReadBytes( BBDataBuffer *buffer,int count,int offset );
	int WriteBytes( BBDataBuffer *buffer,int count,int offset );

	bool Open( String path,String mode );
	
private:
	FILE *_file;
	int _off,_len;
};

// ***** filestream.cpp *****

BBFileStream::BBFileStream():_file(0),_off(-1),_len(-1){
}

BBFileStream::~BBFileStream(){
	if( _file ) fclose( _file );
}

bool BBFileStream::Open( String path,String mode ){

	if( _file ) return false;
	
#if _WIN32	
	_file=_wfopen( path.ToCString<wchar_t>(),mode.ToCString<wchar_t>() );
#else
	_file=fopen( path.ToUTF8(),mode.ToUTF8() );
#endif

	if( !_file ) return false;
	
	fseek( _file,0,SEEK_END );
	_off=0;
	_len=ftell( _file );
	fseek( _file,0,SEEK_SET );
	
	return true;
}

int BBFileStream::Length(){
	return _len;
}

int BBFileStream::Offset(){
	return _off;
}

int BBFileStream::Seek( int offset ){
	if( !_file ) return -1;
	fseek( _file,0,SEEK_SET );
	_off=ftell( _file );
	return _off;
}

int BBFileStream::Eof(){
	return _file ? _off==_len : -1;
}

void BBFileStream::Close(){
	if( !_file ) return;
	fclose( _file );
	_file=0;
	_off=-1;
	_len=-1;
}

int BBFileStream::SkipBytes( int count ){
	if( !_file ) return 0;
	int n=0;
	char buf[1024];
	while( count ){
		int c=count<1024 ? count : 1024;
		int t=fread( buf,1,c,_file );
		count-=t;
		n+=t;
		if( t!=c ) break;
	}
	_off+=n;
	return n;
}

int BBFileStream::ReadBytes( BBDataBuffer *buffer,int count,int offset ){
	if( !_file ) return 0;
	int n=fread( buffer->WritePointer(offset),1,count,_file );
	_off+=n;
	return n;
}

int BBFileStream::WriteBytes( BBDataBuffer *buffer,int count,int offset ){
	if( !_file ) return 0;
	int n=fwrite( buffer->ReadPointer(offset),1,count,_file );
	_off+=n;
	if( _off>_len ) _len=_off;
	return n;
}
